contracts/modules/Tournament.sol

```diff
@@ -0,0 +1,421 @@
+        if (_startTime < t.startTime || _endTime > t.endTime) {
+            revert InvalidEventPeriod();
+        }
+        uint256 eventId = t.nextEventId;
+        ts.tournamentEvents[_id][eventId] = ITournament.TournamentEvent({
+            id: eventId,
+            startTime: _startTime,
+            endTime: _endTime
+        });
+
+        t.nextEventId++;
+
+        emit NewTournamentEvent(_id, eventId, _startTime, _endTime);
+    }
+
+    function updateEvent(
+        bytes32 _id,
+        uint256 _eventId,
+        uint256 _startTime,
+        uint256 _endTime
+    )
+        external
+        virtual
+        override
+        tournamentExists(_id)
+        tournamentNotBanned(_id)
+        tournamentOwnerOrAdmin(_id)
+        validPeriod(_startTime, _endTime)
+    {
+        TournamentStore storage ts = TournamentStorage.load();
+        ITournament.Tournament storage t = ts.tournaments[_id];
+        ITournament.TournamentEvent storage events = ts.tournamentEvents[_id][
+            _eventId
+        ];
+        if (_startTime < t.startTime || _endTime > t.endTime) {
+            revert InvalidEventPeriod();
+        }
+        if (events.id != _eventId) {
+            revert TournamentEventNotFound();
+        }
+        events.startTime = _startTime;
+        events.endTime = _endTime;
+        emit TournamentEventUpdated(_id, _eventId, _startTime, _endTime);
+    }
+
+    function banTournament(
+        bytes32 _id
+    )
+        external
+        virtual
+        override
+        tournamentExists(_id)
+        tournamentNotBanned(_id)
+        onlySoccersmCouncil
+    {
+        TournamentStore storage ts = TournamentStorage.load();
+        ITournament.Tournament storage t = ts.tournaments[_id];
+        t.banned = true;
+        emit TournamentBanned(_id, msg.sender, true);
+    }
+
+    function unBanTournament(
+        bytes32 _id
+    ) external virtual override tournamentExists(_id) onlySoccersmCouncil {
+        TournamentStore storage ts = TournamentStorage.load();
+        ITournament.Tournament storage t = ts.tournaments[_id];
+        require(t.banned, "Tournament not banned");
+        t.banned = false;
+        emit TournamentUnbanned(_id, msg.sender, false);
+    }
+
+    function setTournamentWinner(
+        bytes32 _id,
+        address _winner
+    )
+        external
+        virtual
+        override
+        tournamentNotBanned(_id)
+        tournamentExists(_id)
+        tournamentOwnerOrAdmin(_id)
+        positiveAddress(_winner)
+    {
+        TournamentStore storage ts = TournamentStorage.load();
+        ITournament.Tournament storage t = ts.tournaments[_id];
+        if (block.timestamp < t.endTime) {
+            revert TournamentStillOngoing();
+        }
+        if (!ts.isPlayer[_id][_winner]) {
+            revert NotTournamentPlayer();
+        }
+        t.winner = _winner;
+        emit TournamentWinnerSet(_id, _winner);
+    }
```

Comment:
This function should be removed. Tournament winner is per event. Should be replaced with a function that allows setting winner per event.